feat: New app install flow 2

ibex088 opened the pull request:
## What does this PR do?



Take over from @ThyMinimalDev existing work and complete the following tasks

* When clicking on install new app button, only redirect to the new app install flow if it is an oAuth app or an app that extends the EventType feature in it's config

* adapt oauth apps api callback / add to redirect to new app install flow

ibex088 committed:
fix: redirect to event-type after oauth flow

ibex088 committed:
When clicking on install new app button, only redirect to the new app install flow if it is an oAuth app or an app that extends the EventType feature in it's config

ibex088 committed:
code refactor, skip event-type step

ibex088 committed:
removed extra consoles

ibex088 committed:
so that no one can change the eventTypeId in url

joeauyeung commented on packages/app-store/_components/OmniInstallAppButton.tsx:
```diff
@@ -21,11 +24,13 @@ export default function OmniInstallAppButton({
   className: string;
   returnTo?: string;
   teamId?: number;
+  eventType: EventTypeAppSettingsComponentProps["eventType"];
```
I notice you're passing the whole event type object but we're only using the event type id. I think we should just pass the id down.

joeauyeung commented on packages/app-store/_components/AppCard.tsx:
```diff
@@ -34,6 +35,7 @@ export default function AppCard({
   disableSwitch?: boolean;
   switchTooltip?: string;
   hideSettingsIcon?: boolean;
+  eventType: EventTypeAppSettingsComponentProps["eventType"];
```
See my comment on `OmniInstallAppButton` we should only pass down the event type id.

joeauyeung commented on packages/lib/apps/getAppOnboardingRedirectUrl.ts:
```diff
@@ -0,0 +1,13 @@
+
+import { getAppOnboardingUrl } from "./getAppOnboardingUrl";
+
+export const getAppOnboardingRedirectUrl = (slug: string, teamId?: number, eventTypeId?: number) => {
```
I notice we're using this function across every callback function. Could we instead generate this redirect URL and add it to the OAuth state to be read on at the callback endpoint? 

The reason why is when we're developing new apps it's extra mental overhead to remember to add the `getAppOnboardingRedirectUrl` function to every callback endpoint. Whenever we can reduce mental overhead, reduces the possibilities of bugs.

joeauyeung reviewed (changes requested):
Great work so far @SomayChauhan. I left a few comments that need to be addressed first. @ThyMinimalDev @sean-brydon can you read through my review comments and let me know if you object to anything. 

## Blocking
- [x]  On app install page it's not respecting dark theme
- [x] On configure step, inputs aren't showing errors (ex. Non empty values for text inputs)
- [x] I don't think calendars should be a part of the flow. They're still only avaialble on a individual scope and not a team scope
- [x] For OAuth flow I don't think we need the OAuth step page. I think once you choose the account to install the app on it should immedietly take you to the app provider's OAuth page
- [x] When coming back from the OAuth provider's page, it redirects me back to the App Store. It should redirect back to choosing which event type to enable Stripe on and the config
    - [x] I just tested this again for a team and it brought me to the event type selection page. Although when I hit next it didn't bring me to the config step but back to the app store

## Non-blocking
- [x] Should be able to choose multiple event types
- [x] Not a huge fan that the new flow is named "onboarding". It might cause some confusion with our actual onboarding wizard. Can we maybe call the pathname "/apps/installation/"

![image](https://github.com/calcom/cal.com/assets/65426560/5f03644f-7aef-4cae-995f-dba619ccfb89)

ibex088 committed:
theme support enabled

ibex088 committed:
only pass eventTypeId instead of whole evetType obj to OmniInstallAppButton

ibex088 committed:
fix: clicking on event-type gives error and redirects to /apps

ibex088 committed:
saved appOnboardingRedirectUrl to OAuth State

ibex088 committed:
fix: a user having no team breaks the onboarding flow

ibex088 committed:
add optional chaining to appMetadata.extendsFeature and appMetadata.isOAuth;

ibex088 committed:
encode appOnboardingRedirectUrl

ibex088 committed:
fix: dark theme ui issues

SomayChauhan commented on apps/web/pages/apps/onboarding/[[...step]].tsx:
```diff
@@ -255,14 +255,15 @@ const OnboardingPage = ({
-          "--cal-brand-subtle": "#9CA3AF",
-        } as CSSProperties
-      }>
+      // style={
```
not sure why it was added, but this is causing some dark mode ui issues

ibex088 committed:
skip configure step if no-event types

if user/team does not have event-type then skip configure step and goto apps/installed/[category]

SomayChauhan commented on apps/web/pages/apps/onboarding/[[...step]].tsx:
```diff
@@ -512,15 +512,20 @@ export const getServerSideProps = async (context: GetServerSidePropsContext) =>
+        const _eventTypes = await getEventTypes(user.id, parsedTeamIdParam);
+
+        if (!_eventTypes) {
+          return { redirect: { permanent: false, destination: `/apps/installed/${appMetadata.category}` } };
```
if user/team does not have event-type then skip configure step and goto apps/installed/[category]

ibex088 committed:
exclude calendars from apps- onboarding flow

ibex088 committed:
Update getAppOnboardingRedirectUrl.ts

SomayChauhan commented:
- [ ] edge case: multiple payment apps for a event type

ibex088 committed:
fix: On configure step, inputs aren't showing errors

ibex088 committed:
change route from `apps/onboarding` to `apps/installation`

ibex088 committed:
fix: flow not correctly working for teams

ibex088 committed:
should only allow team ADMIN or OWER to install an app to team

ibex088 committed:
Prevent two payment apps to be enabled

ibex088 committed:
removed loading state

joeauyeung committed:
Merge branch 'new-app-install-flow' into new-app-install-flow-2

joeauyeung reviewed (changes requested):
Great work so far @SomayChauhan. There are still some points that need to be addressed.

## Blocking
- [x] Not seeing an error message on the installation flow. For example, I'm installing GTM and when I leave "TrackingId" blank I see an error on the backend but not on the frontend.
- [x] I still don't think we need this setup page for OAuth apps. After selecting which account to install it on it should just bring you to the 3rd party OAuth page
![image](https://github.com/calcom/cal.com/assets/65426560/58910a5e-a21c-404d-adaa-15e988da4905)
- [x] On the install flow, it's starting at step 0 instead of step 1
![image](https://github.com/calcom/cal.com/assets/65426560/d48227e3-3e58-4ce5-8cbb-a8d87fab1b16)
- [ ] For stripe, after OAuth install and choosing an event type, when saving the data I'm getting an error on the backend that the `credentialId` isn't being passed. For payment apps and any event type app that requires a `credentialId` we need to ensure that this data is being passed


## Non-blocking
- [x] @ThyMinimalDev @sean-brydon I'm not a huge fan of how when installing an app via the Apps tab on the event type page it jumps you to the installation wizard. Seems pretty jarring being jumped around screens like that. For OAuth though I can understand.
https://www.loom.com/share/1d82fdda4f4b4f9bb1b13c3c2195e7ab

joeauyeung committed:
Fix type error

joeauyeung commented on apps/web/components/apps/installation/ConfigureStepCard.tsx:
```diff
@@ -28,18 +28,22 @@ export const ConfigureStepCard: FC<ConfigureStepCardProps> = ({
   const { getAppDataGetter, getAppDataSetter } = useAppsData();
-  const { shouldLockDisableProps } = useLockedFieldsManager(
+  const formMethods = useForm();
+  const { shouldLockDisableProps } = useLockedFieldsManager({
```
@ThyMinimalDev @sean-brydon, do we need the useLockedFieldsManager in the app installation flow? I see that apps are locked as a whole for managed event types and not on the app level. This is testing with GTM. I think if we pass the right data through here we don't need to use this function and worry about form methods. The formMethds taht useLockedFieldsManager expects is for a whole event type.

ibex088 committed:
undo new-app-flow from OmniInstallAppButton

ibex088 committed:
remove OAuth Step, handle step numbering

ibex088 committed:
fix: redirect url

ibex088 committed:
getAppOnboardingRedirectUrl refactor

ibex088 committed:
for debugging

ibex088 committed:
Revert "for debugging"

This reverts commit b9362c1a8f47f2280e684c213d7ec08d04a8ac4c.

ibex088 committed:
fix: added toaster

ibex088 committed:
feat: added support to select multiple events (almost)

SomayChauhan commented on apps/web/components/apps/installation/ConfigureStepCard.tsx:
```diff
@@ -1,61 +1,158 @@
+        className="text-md mt-6 w-full justify-center"
+        onClick={() => {
+          // submitRefs.current[0].current?.click();
+          submitRefs.current.map((ref) => ref.current?.click());
```
@joeauyeung 
is there a better way to handle this?
its getting kinda messy?


the flow so far - 
[LOOM](https://www.loom.com/share/d6e517832a5c47cf8fdf313ba165e0c1?sid=b3873b0d-d50a-4ac9-a814-e50be0bf6602)

joeauyeung replied:
I think this is awesome. Allowing users to configure multiple event types is going to help a ton. This gets my seal of approval. Thoughts @ThyMinimalDev @sean-brydon @ciaranha @itsberkaya?

ibex088 committed:
hide button

ibex088 committed:
Merge branch 'new-app-install-flow' into new-app-install-flow-2

ibex088 committed:
Move inner form outside outer form using portal

Due to limitations with nested forms in React Hook Form, moved the inner form outside the outer form using a portal to ensure proper form handling.

ibex088 committed:
submit buttons refs madness

ibex088 committed:
removed configure step

we had to use query params to manage selected event-types, between routes which was causing 2 issues:
* each time we selected or unselected an event the whole page was reloaded including getserversideprops being called again and again
* the form was reloaded causing more issues

ibex088 committed:
removed extra unwanted code

ibex088 committed:
small fix

SomayChauhan commented:
https://www.loom.com/share/85fa766b64e043fe864c29551305b365?sid=e38f657d-e08f-43b5-8a02-4105860f0160

ibex088 committed:
fix: type error

ibex088 committed:
fix: types

joeauyeung commented on apps/web/components/apps/installation/ConfigureStepCard.tsx:
```diff
@@ -0,0 +1,211 @@
+          })}
+        </div>
+
+        <button form="outer-event-type-form" type="submit" className="hidden" ref={mainForSubmitRef}>
+          Save
+        </button>
+        <Button
+          className="text-md mt-6 w-full justify-center"
+          type="button"
+          onClick={(e) => {
+            submitRefs.current.reverse().map((ref) => ref.current?.click());
+            setSubmit(true);
+          }}
+          loading={loading}>
+          Save
+        </Button>
```
Can you explain what these two buttons are for? Why is one of them hidden?

SomayChauhan replied:
here in this scenario we have nested forms, and we want to submit the outer form only when all the inner forms have been filled out.
in this case these are the 2 individual inner forms
![Screenshot 2024-04-03 at 6 17 09 AM](https://github.com/calcom/cal.com/assets/32706411/038a9c56-46ee-41a3-9a4f-55e2a6172dd8)

`submitRefs.current.reverse().map((ref) => ref.current?.click());` is iterating over the inner forms and triggering their submission. However, we want to ensure that the outer form is only submitted when all the inner forms contain values. To achieve this, you need to keep track of whether all the inner forms are filled out. this is what `updatedEventTypesStatus` state is for.

the hidden button is the actual submit button for the outer form which we trigger when all the inner form values have been filled out and saved.

joeauyeung commented on apps/web/components/apps/installation/StepFooter.tsx:
Do we need this component in `[[...step]].tsx`? Commenting this out I didn't see a difference in UI
![image](https://github.com/calcom/cal.com/assets/65426560/7b06ee9b-dd51-4742-b2be-d51b940334cf)

If we're not adding a "Go Back" button we should also remove the commented code.

joeauyeung commented on apps/web/pages/apps/installation/[[...step]].tsx:
```diff
@@ -0,0 +1,509 @@
+      }
+
+      if (err.data?.code === "UNAUTHORIZED") {
+        message = `${err.data.code}: ${t("error_event_type_unauthorized_update")}`;
+        message = `${err.data.code}: ${t(err.message)}`;
```
Message is being assigned two different values back to back. Which one should we use?

joeauyeung commented on apps/web/pages/apps/installation/[[...step]].tsx:
```diff
@@ -0,0 +1,509 @@
+      key={pathname}
+      className="dark:bg-brand dark:text-brand-contrast text-emphasis min-h-screen px-4"
+      data-testid="onboarding"
+      // style={
+      //   {
+      //     "--cal-brand": "#111827",
+      //     "--cal-brand-emphasis": "#101010",
+      //     "--cal-brand-text": "white",
+      //     "--cal-brand-subtle": "#9CA3AF",
+      //   } as CSSProperties
+      // }
```
```suggestion

```

joeauyeung commented on apps/web/pages/apps/installation/[[...step]].tsx:
```diff
@@ -0,0 +1,509 @@
+      setIsSelectingAccount(true);
+      if (appMetadata.isOAuth) {
+        const state = JSON.stringify({
+          appOnbaordingRedirectUrl: getAppOnboardingRedirectUrl(appMetadata.slug, teamId),
```
```suggestion
          appOnboardingRedirectUrl: getAppOnboardingRedirectUrl(appMetadata.slug, teamId),
```
When fixing the typo here. Ensure that other parts of the app aren't expecting the typo'd value

joeauyeung commented on apps/web/pages/apps/installation/[[...step]].tsx:
```diff
@@ -0,0 +1,509 @@
+      // }
+    >
+      <Head>
+        <title>Install {appMetadata?.name ?? ""}</title>
```
Let's add a translation here

joeauyeung commented on packages/app-store/_utils/oauth/decodeOAuthState.ts:
```diff
@@ -7,6 +7,9 @@ export function decodeOAuthState(req: NextApiRequest) {
     return undefined;
   }
   const state: IntegrationOAuthCallbackState = JSON.parse(req.query.state);
+  if (state.appOnbaordingRedirectUrl) {
+    state.appOnbaordingRedirectUrl = decodeURIComponent(state.appOnbaordingRedirectUrl);
+  }
```
Let's fix the typos here as mentioned earlier

joeauyeung reviewed (changes requested):
Great job so far @SomayChauhan feels like we're getting closer to getting this merged.

## Blocking
- I left some comments about a type in `state.appOnbaordingRedirectUrl` which should be `state.appOnboardingRedirectUrl`. Let's clean that up across different files.
- When selecting to install an app on a team
	- It highlights both the team and the user on the accounts step
	- On the event types step, even when a team is selected, the user's personal event types appear and no the team's
	- https://www.loom.com/share/9241b5dc2aad43b5b3fbc84f120eca70

ibex088 committed:
implemented suggested changes

* removed step footer
* fixed error.message
* removed commented styles

ibex088 committed:
removed step footer

ibex088 committed:
fix: Display team event types instead of user's on event types step when team selected

ibex088 committed:
fix: typo

ibex088 committed:
removed shouldLockDisableProps

ibex088 committed:
refactor setUpdatedEventTypesStatus, fix type

ibex088 committed:
removed extra code

ibex088 committed:
can't delete while saving

ibex088 committed:
added translations

ibex088 committed:
fix: loading in select account step, fix handle onSelect params type

SomayChauhan commented on apps/web/components/apps/installation/ConfigureStepCard.tsx:
```diff
@@ -0,0 +1,211 @@
+  return (
+    <EventTypeAppSettings
+      slug={slug}
+      disabled={shouldLockDisableProps("apps").disabled}
```
yes, removed 💯

ibex088 committed:
Merge branch 'new-app-install-flow' into new-app-install-flow-2

joeauyeung committed:
Fix typo

joeauyeung reviewed (approved):
Went through everything and it looks good to me! Great job @SomayChauhan!

Let's chat with @ThyMinimalDev and @sean-brydon to see what's the best way to get this merged.
